lib/mapObjects/Rewardable.h

```diff
@@ -0,0 +1,350 @@
+		h & manaDiff;
+		h & manaOverflowFactor;
+		h & movePoints;
+		h & primary;
+		h & secondary;
+		h & bonuses;
+		h & artifacts;
+		h & spells;
+		h & creatures;
+		h & creaturesChange;
+		if(version >= 821)
+			h & spellCast;
+	}
+};
+
+class DLL_LINKAGE CRewardVisitInfo
+{
+public:
+	enum ERewardEventType
+	{
+		EVENT_INVALID,
+		EVENT_FIRST_VISIT,
+		EVENT_ALREADY_VISITED,
+		EVENT_NOT_AVAILABLE
+	};
+
+	CRewardLimiter limiter;
+	CRewardInfo reward;
+
+	/// Message that will be displayed on granting of this reward, if not empty
+	MetaString message;
+
+	/// Event to which this reward is assigned
+	ERewardEventType visitType;
+
+	CRewardVisitInfo() = default;
+
+	template <typename Handler> void serialize(Handler &h, const int version)
+	{
+		h & limiter;
+		h & reward;
+		h & message;
+		h & visitType;
+	}
+};
+
+namespace Rewardable
+{
+	enum EVisitMode
+	{
+		VISIT_UNLIMITED, // any number of times. Side effect - object hover text won't contain visited/not visited text
+		VISIT_ONCE,      // only once, first to visit get all the rewards
+		VISIT_HERO,      // every hero can visit object once
+		VISIT_BONUS,     // can be visited by any hero that don't have bonus from this object
+		VISIT_PLAYER     // every player can visit object once
+	};
+
+	/// controls selection of reward granted to player
+	enum ESelectMode
+	{
+		SELECT_FIRST,  // first reward that matches limiters
+		SELECT_PLAYER, // player can select from all allowed rewards
+		SELECT_RANDOM, // one random reward from all mathing limiters
+	};
+
+	const std::array<std::string, 3> SelectModeString{"selectFirst", "selectPlayer", "selectRandom"};
+	const std::array<std::string, 5> VisitModeString{"unlimited", "once", "hero", "bonus", "player"};
+
+	/// Base class that can handle granting rewards to visiting heroes.
+	struct DLL_LINKAGE Configuration
+	{
+		/// Message that will be shown if player needs to select one of multiple rewards
+		MetaString onSelect;
+
+		/// Rewards that can be applied by an object
+		std::vector<CRewardVisitInfo> info;
+
+		/// how reward will be selected, uses ESelectMode enum
+		ui8 selectMode = Rewardable::SELECT_FIRST;
+
+		/// contols who can visit an object, uses EVisitMode enum
+		ui8 visitMode = Rewardable::VISIT_UNLIMITED;
+
+		/// how and when should the object be reset
+		CRewardResetInfo resetParameters;
+
+		/// if true - player can refuse visiting an object (e.g. Tomb)
+		bool canRefuse = false;
+
+		/// if true - object info will shown in infobox (like resource pickup)
+		EInfoWindowMode infoWindowType = EInfoWindowMode::AUTO;
+		
+		EVisitMode getVisitMode() const;
+		ui16 getResetDuration() const;
+		
+		template <typename Handler> void serialize(Handler &h, const int version)
+		{
+			h & info;
+			h & canRefuse;
+			h & resetParameters;
+			h & onSelect;
+			h & visitMode;
+			h & selectMode;
+			h & infoWindowType;
+		}
+	};
+
+	class DLL_LINKAGE Interface
+	{
+	private:
+		
+		Rewardable::Configuration _configuration;
```

Comment:
1. We don't use underscores for member names
2. Might be better to use created in consturctor unique_ptr and move interface class into a separate file to minimize includes. This header already includes quite a lot and this file is likely to be included in many other files in the future.

Reply:
as I understood, we cannot use forward declaration if serialization method is presented

Reply:
We can, but you need to add `#include` in cpp that performs template instantiations (and only that file)

Can't remember which one, check compiler output to find what file triggered build error.

Reply:
@IvanSavenko agree, we can. I splitting HeroBonus.cpp by your request now and faced same issues - `#include` in around 8 places fixed it.